design/DFX-Interface.md

```diff
@@ -99,22 +91,27 @@ This file informs Motoko about the interface of that canister. It could be the o
 
 Open problem: how to resolve mutual canister imports.
 
-Compiling IDL Files to JS
--------------------------
-
-In order to compile an IDL file, `dfx` invokes `didc` with
-
-    didc --js some/path/input.did -o another/path/output.js
-
-This _reads_ `some/path/input.did` and any `.did` file referenced by
-`some/path/input.did`.
-
-No constraints are imposed where these imported files reside (this may be refined to prevent relative imports from looking outside the project and the declared packages)
-
-This _writes_ to `another/path/output.js`, but has no other effect. It does
-not create `another/path/`.
-
-Compiler warnings and errors are reported to `stderr`. Nothing writes to `stdout`.
+Exporting Canister Metadata
+---------------------------
+
+The compiler generates various metadata about the canister via command line flags.
+The compiled Wasm module also includes the metadata in the custom sections.
+
+* Candid interface.
+  + Compiler flag `--idl` generates the Candid interface for the canister. The main service
+    is always a service constructor, which contains the initialization arguments for installing the canister.
+  + Custom section `icp:public candid` stores the interface for the running (initialized) canister, which removes
+    the initializatio arguments. This custom section is publicly accessible.
+  + Custom section `icp:private init_args` stores the initialization arguments. The argument types can refer to
+    types defined in the `icp:public candid` custom section. This custom section is only accessible by the controllers
+    of the canister.
```

Comment:
Yes, we already merged the public spec part. Replica implementation is in progress. 

We only expose custom sections with prefix `icp:`. @rossberg insists exposing custom sections be a conscious decision. If developers want to expose a non-IC-specific metadata, they need to explicitly declare that in the custom section name.

Reply:
My reasoning is that non-icp custom sections are normally implementation details of the wasm code, and it does not make sense to access them without the code itself. You really only want to publish metadata on the IC that is intended and designed for that purpose.

Reply:
Allowing public access to the binary itself would be a useful mechanism as well, I agree. But that's different from metadata. We could permit it e.g. by having the IC interpret the presence of an (empty) section `icp:public-wasm` or something like that.

> Also, I don’t understand the “private” designation. Isn’t every section (custom or not) private?

The metadata is public or private, not the section encoding it. Thinking about this feature in terms of "exposing custom sections" is conflating functionality and implementation. It is a feature for attaching public or private metadata to an IC canister. The fact that it is encoded in Wasm custom sections, and how exactly, is a lower-level implementation detail. In the future, we could add other ways of producing it.

Reply:
Are you saying that the system-level specification (the stuff written in the interface spec) will actually parse and understand the content of these sections, instead of just providing a generic mechanism to expose the raw data therein?

Reply:
Failure if we’d support lua and x64 in particular? Or _anything_ that isn’t Wasm? Why would that be a failure of the IC?

Reply:
Ok, then it would be a failure of Wasm if there were a genuine demand for such canisters  – but if that happens and the IC’s design would seamlessly allow that, then I’m not sure I’d count that as failure, but rather as fortunate foresight – and a prime example why good implementation-hidings abstractions, such as the one you defended here, are useful.

Anyways, this is getting too philosophical for a Github PR comment thread :-)